src/librustdoc/html/sources.rs

```diff
@@ -195,32 +199,33 @@ impl SourceCollector<'_, '_> {
             title: &title,
             css_class: "source",
             root_path: &root_path,
-            static_root_path: self.cx.shared.static_root_path.as_deref(),
+            static_root_path: shared.static_root_path.as_deref(),
             description: &desc,
             keywords: BASIC_KEYWORDS,
-            resource_suffix: &self.cx.shared.resource_suffix,
-            extra_scripts: &[&format!("source-files{}", self.cx.shared.resource_suffix)],
-            static_extra_scripts: &[&format!("source-script{}", self.cx.shared.resource_suffix)],
+            resource_suffix: &shared.resource_suffix,
+            extra_scripts: &[&format!("source-files{}", shared.resource_suffix)],
+            static_extra_scripts: &[&format!("source-script{}", shared.resource_suffix)],
         };
         let v = layout::render(
-            &self.cx.shared.layout,
+            &mut self.cx,
+            &shared.layout,
             &page,
             "",
-            |buf: &mut _| {
+            |buf: &mut _, cx: &mut Context<'_>| {
```

Comment:
I suspect the reason the code is slower is because you're now passing `cx` to all print functions. Can you instead call print_src ahead of time and pass that string to render()? That would allow you to remove the cx parameter in the Print trait.

Reply:
I think it'd be better indeed. Let's give it a try.